processor.go

```diff
@@ -49,335 +49,356 @@ type KickRequest[JC any] struct {
+	}
+	return ret
+}
+
+// Serializer is an interface that defines how to serialize and deserialize job contexts.
+
 // Processor executes a job
 type Processor[AC any, OC any, JC any] struct {
 	appContext     AC
-	states         []State[AC, OC, JC]
 	serializer     Serializer[OC, JC]
+	stateThing     stateThing[AC, OC, JC]
 	statusListener StatusListener
-	initted        bool
-	stateMap       map[string]State[AC, OC, JC]
-	stateNames     []string
-	stateChan      map[string]chan Job[JC]
 	returnChan     chan Return[JC]
-}
-
-func NewProcessor[AC any, OC any, JC any](ac AC, states []State[AC, OC, JC], serializer Serializer[OC, JC], statusListener StatusListener) *Processor[AC, OC, JC] {
-	return &Processor[AC, OC, JC]{
-		appContext:     ac,
-		states:         states,
-		serializer:     serializer,
-		statusListener: statusListener,
-	}
+	wg             sync.WaitGroup
 }
 
 // Return is a struct that contains a job and a list of kick requests
 // that is used for returning job updates to the system
 type Return[JC any] struct {
+	PriorState   string
 	Job          Job[JC]
 	KickRequests []KickRequest[JC]
 	Error        error
 }
 
-func (p *Processor[AC, OC, JC]) init() {
-	if p.initted {
-		return
+func NewProcessor[AC any, OC any, JC any](ac AC, states []State[AC, OC, JC], serializer Serializer[OC, JC], statusListener StatusListener) *Processor[AC, OC, JC] {
+	return &Processor[AC, OC, JC]{
+		appContext:     ac,
+		stateThing:     newStateThingFromStates(states),
+		serializer:     serializer,
+		statusListener: statusListener,
 	}
+}
+
+func (p *Processor[AC, OC, JC]) init() {
 	if p.serializer == nil {
 		p.serializer = &NilSerializer[OC, JC]{}
 	}
 	if p.statusListener == nil {
 		p.statusListener = &NilStatusListener{}
 	}
-	// Make a map of triggers to states so we can easily reference it
-	p.stateMap = map[string]State[AC, OC, JC]{}
-	for _, s := range p.states {
-		p.stateMap[s.TriggerState] = s
-	}
-	// get a list of return state names for use
-	p.stateNames = make([]string, 0, len(p.stateMap))
-	for k := range p.stateMap {
-		p.stateNames = append(p.stateNames, k)
-	}
-
-	// For each state, we need a channel of jobs
-	p.stateChan = map[string]chan Job[JC]{}
-
-	// Create the state chans
-	totalConcurrency := 0
-	for _, s := range p.states {
-		if s.Terminal {
-			continue
-		}
-		p.stateChan[s.TriggerState] = make(chan Job[JC], s.Concurrency) // make a chan
-		totalConcurrency += s.Concurrency
-	}
 
-	// When a job changes state, we send it to this channel to centrally manage and re-queue
-	p.returnChan = make(chan Return[JC], totalConcurrency*2) // make it the size of the total amount of in flight jobs we could have so that each worker can return a task
+	// This is by-design unbuffered
+	p.returnChan = make(chan Return[JC])
 }
 
 // Exec this big work function, this does all the crunching
 func (p *Processor[AC, OC, JC]) Exec(ctx context.Context, r *Run[OC, JC]) error {
+	if err := p.stateThing.validate(); err != nil {
```

Comment:
should we make NewProcessor return an error, it is a breaking change but that's the earliest time we would know the states are invalid